File: backend/app/models/event.py
"""Event model — created by a customer to plan a Marriage/Birthday/Corporate event."""
import enum
from datetime import date, datetime

from sqlalchemy import Date, DateTime, Enum, Float, ForeignKey, Integer, String, func
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.db.base import Base


class EventType(str, enum.Enum):
    MARRIAGE = "marriage"
    BIRTHDAY = "birthday"
    CORPORATE = "corporate"


class Event(Base):
    __tablename__ = "events"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, index=True)
    user_id: Mapped[int] = mapped_column(ForeignKey("users.id", ondelete="CASCADE"), nullable=False)

    name: Mapped[str] = mapped_column(String(150), nullable=False)
    event_type: Mapped[EventType] = mapped_column(Enum(EventType, name="event_type"), nullable=False)
    event_date: Mapped[date] = mapped_column(Date, nullable=False)
    location: Mapped[str] = mapped_column(String(255), nullable=False)
    total_budget: Mapped[float] = mapped_column(Float, nullable=False)

    created_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now())
    updated_at: Mapped[datetime] = mapped_column(DateTime(timezone=True), server_default=func.now(), onupdate=func.now())

    user: Mapped["User"] = relationship("User", back_populates="events")
    budget_allocations: Mapped[list["BudgetAllocation"]] = relationship(
        "BudgetAllocation", back_populates="event", cascade="all, delete-orphan", order_by="BudgetAllocation.sort_order"
    )
    bookings: Mapped[list["Booking"]] = relationship("Booking", back_populates="event", cascade="all, delete-orphan")

File: frontend/src/types/event.ts
export type EventType = "marriage" | "birthday" | "corporate";

export interface EventItem {
  id: number;
  user_id: number;
  name: string;
  event_type: EventType;
  event_date: string;
  location: string;
  total_budget: number;
  created_at: string;
  updated_at: string;
}

export interface BudgetAllocation {
  id: number;
  event_id: number;
  name: string;
  icon: string;
  allocated_amount: number;
  spent_amount: number;
  sort_order: number;
  updated_at: string;
}

export interface BudgetSummary {
  total_budget: number;
  total_allocated: number;
  total_spent: number;
  remaining: number;
  is_over_budget: boolean;
  categories: BudgetAllocation[];
}
